pkg/operator/ceph/controller/cleanup.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2024 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controller
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	batch "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	k8sError "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	volumeName                  = "cleanup-volume"
+	dataDirHostPath             = "ROOK_DATA_DIR_HOST_PATH"
+	CleanupAppName              = "resource-cleanup"
+	RESOURCE_CLEANUP_ANNOTATION = "ceph.rook.io/force-deletion"
+
+	// CephFSSubVolumeGroup env resources
+	CephFSSubVolumeGroupNameEnv = "SUB_VOLUME_GROUP_NAME"
+	CephFSVolumeNameEnv         = "VOLUME_NAME"
+	CephFSNamesaceEnv           = "NAMESPACE"
+	CephFSMetaDataPoolNameEnv   = "METADATA_POOL_NAME"
+)
+
+// ResourceCleanup defines an rook ceph resource to be cleaned up
+type ResourceCleanup struct {
+	resource  k8sClient.Object
+	cluster   *cephv1.CephCluster
+	rookImage string
+	// config defines the attributes of the custom resource to passed in as environment variables in the clean up job
+	config map[string]string
+}
+
+func NewResourceCleanup(obj k8sClient.Object, cluster *cephv1.CephCluster, rookImage string, config map[string]string) *ResourceCleanup {
+	return &ResourceCleanup{
+		resource:  obj,
+		rookImage: rookImage,
+		cluster:   cluster,
+		config:    config,
+	}
+}
+
+// Start a new job to perform clean up of the ceph resources. It returns true if the cleanup job has succeeded
+func (c *ResourceCleanup) Start(ctx context.Context, client k8sClient.Client) (bool, error) {
+	jobName := k8sutil.TruncateNodeNameForJob("cleanup-job-%s", c.resource.GetName())
+	podSpec := c.jobTemplateSpec()
+	job := &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            jobName,
+			Namespace:       c.resource.GetNamespace(),
+			OwnerReferences: c.resource.GetOwnerReferences(),
+		},
+		Spec: batch.JobSpec{
+			Template: podSpec,
+		},
+	}
+
+	existingJob := &batch.Job{}
```

Comment:
Before we get or create the job, let's check if there are any subvolumes in the subvolumegroup. if not, then we don't need to create the job at all in the first place and it can be immediately deleted. Or is that check somewhere else in the finalizer already?

Reply:
Currently are always creating a job for cleaning up the subvolumegroup, whether it has  subvolumes or not.

Reply:
That looks like good to have. But since we only create this job when annotation is present on the resource, it should be ok to run the job as user has deliberately asked to clean the resource (subvolumes, snapshot etc)

Reply:
If there are no subvolumes in the svg, seems like the finalizer should be immediately removed. This would cover both cases of 1) no subvolumes existing (and no need for creating the job), or 2) if the subvolumes were already cleaned up by the job and the operator is retrying to delete the svg CR.